File: core/state-cache.js
// Bucu Core - State Cache (JavaScript)
// Lightweight in-memory key-value store with TTL support

const StateCache = {
    _cache: new Map(),  // Map<string, { value: any, expiry: number | null }>
    _stats: {
        hits: 0,
        misses: 0,
        sets: 0,
        deletes: 0
    },
    
    // Set a value in cache with optional TTL
    set(key, value, ttl) {
        // Validate key
        if (typeof key !== "string" || key === "") {
            if (typeof Logger !== 'undefined') {
                Logger.error("Cache key must be a non-empty string");
            }
            return false;
        }
        
        // Calculate expiry time
        let expiry = null;
        if (ttl && typeof ttl === "number" && ttl > 0) {
            expiry = Date.now() + (ttl * 1000);  // Convert seconds to milliseconds
        }
        
        // Store in cache
        this._cache.set(key, {
            value: value,
            expiry: expiry
        });
        
        this._stats.sets++;
        return true;
    },
    
    // Get a value from cache
    get(key) {
        // Validate key
        if (typeof key !== "string" || key === "") {
            return null;
        }
        
        const entry = this._cache.get(key);
        
        // Check if entry exists
        if (!entry) {
            this._stats.misses++;
            return null;
        }
        
        // Check if entry has expired
        if (entry.expiry && Date.now() > entry.expiry) {
            this._cache.delete(key);
            this._stats.misses++;
            return null;
        }
        
        this._stats.hits++;
        return entry.value;
    },
    
    // Delete a value from cache
    delete(key) {
        if (typeof key !== "string" || key === "") {
            return false;
        }
        
        if (this._cache.has(key)) {
            this._cache.delete(key);
            this._stats.deletes++;
            return true;
        }
        
        return false;
    },
    
    // Check if a key exists in cache
    has(key) {
        return this.get(key) !== null;
    },
    
    // Clear all cache entries
    clear() {
        const count = this._cache.size;
        this._cache.clear();
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Cache cleared: ${count} entries removed`);
        }
        
        return count;
    },
    
    // Clear expired entries
    clearExpired() {
        let count = 0;
        const now = Date.now();
        
        for (const [key, entry] of this._cache.entries()) {
            if (entry.expiry && now > entry.expiry) {
                this._cache.delete(key);
                count++;
            }
        }
        
        if (count > 0 && typeof Logger !== 'undefined') {
            Logger.debug(`Cleared ${count} expired cache entries`);
        }
        
        return count;
    },
    
    // Get cache statistics
    getStats() {
        const totalRequests = this._stats.hits + this._stats.misses;
        const hitRate = totalRequests > 0 
            ? (this._stats.hits / totalRequests * 100) 
            : 0;
        
        return {
            size: this._cache.size,
            hits: this._stats.hits,
            misses: this._stats.misses,
            sets: this._stats.sets,
            deletes: this._stats.deletes,
            hitRate: hitRate
        };
    },
    
    // Get all keys (for debugging)
    keys() {
        return Array.from(this._cache.keys());
    }
};

// Export for Node.js and FiveM
if (typeof module !== 'undefined' && module.exports) {
    module.exports = StateCache;
}

// Make available globally in FiveM
if (typeof global !== 'undefined') {
    global.StateCache = StateCache;
}

File: core/state-cache.lua
-- Bucu Core - State Cache
-- Lightweight in-memory key-value store with TTL support

local StateCache = {
    _cache = {},  -- { [key] = { value, expiry } }
    _stats = {
        hits = 0,
        misses = 0,
        sets = 0,
        deletes = 0
    }
}

-- Set a value in cache with optional TTL
function StateCache:set(key, value, ttl)
    -- Validate key
    if type(key) ~= "string" or key == "" then
        if Logger then
            Logger:error("Cache key must be a non-empty string")
        end
        return false
    end
    
    -- Calculate expiry time
    local expiry = nil
    if ttl and type(ttl) == "number" and ttl > 0 then
        expiry = os.time() + ttl
    end
    
    -- Store in cache
    self._cache[key] = {
        value = value,
        expiry = expiry
    }
    
    self._stats.sets = self._stats.sets + 1
    return true
end

-- Get a value from cache
function StateCache:get(key)
    -- Validate key
    if type(key) ~= "string" or key == "" then
        return nil
    end
    
    local entry = self._cache[key]
    
    -- Check if entry exists
    if not entry then
        self._stats.misses = self._stats.misses + 1
        return nil
    end
    
    -- Check if entry has expired
    if entry.expiry and os.time() > entry.expiry then
        self._cache[key] = nil
        self._stats.misses = self._stats.misses + 1
        return nil
    end
    
    self._stats.hits = self._stats.hits + 1
    return entry.value
end

-- Delete a value from cache
function StateCache:delete(key)
    if type(key) ~= "string" or key == "" then
        return false
    end
    
    if self._cache[key] then
        self._cache[key] = nil
        self._stats.deletes = self._stats.deletes + 1
        return true
    end
    
    return false
end

-- Check if a key exists in cache
function StateCache:has(key)
    return self:get(key) ~= nil
end

-- Clear all cache entries
function StateCache:clear()
    local count = 0
    for _ in pairs(self._cache) do
        count = count + 1
    end
    
    self._cache = {}
    
    if Logger then
        Logger:info("Cache cleared: " .. count .. " entries removed")
    end
    
    return count
end

-- Clear expired entries
function StateCache:clearExpired()
    local count = 0
    local now = os.time()
    
    for key, entry in pairs(self._cache) do
        if entry.expiry and now > entry.expiry then
            self._cache[key] = nil
            count = count + 1
        end
    end
    
    if count > 0 and Logger then
        Logger:debug("Cleared " .. count .. " expired cache entries")
    end
    
    return count
end

-- Get cache statistics
function StateCache:getStats()
    local size = 0
    for _ in pairs(self._cache) do
        size = size + 1
    end
    
    return {
        size = size,
        hits = self._stats.hits,
        misses = self._stats.misses,
        sets = self._stats.sets,
        deletes = self._stats.deletes,
        hitRate = self._stats.hits + self._stats.misses > 0 
            and (self._stats.hits / (self._stats.hits + self._stats.misses) * 100) 
            or 0
    }
end

-- Get all keys (for debugging)
function StateCache:keys()
    local keys = {}
    for key, _ in pairs(self._cache) do
        table.insert(keys, key)
    end
    return keys
end

return StateCache
